# Issue 12137: GLPK crashes or hangs on certain inputs

Issue created by migration from https://trac.sagemath.org/ticket/12309

Original creator: john_perry

Original creation time: 2012-01-15 01:34:37

Assignee: ncohen

CC:  malb ncohen

The attached sage script (`inhomogeneous_bug_test.py`) creates a linear system that

  * crashes Sage if the variables are continuous, and
  * hangs Sage if the variables are integer.

Nathann determined that this was a bug with GLPK, the underlying solver. I have attached a C++ program (`glpk_bug.cpp`) that reproduces the bug directly via the GLPK library, and passed it on to the developers.

I can't resist the temptation to say that the developers don't consider this a bug, but a feature; see

  [This thread on the bug-glpk mailing list archive](http://lists.gnu.org/archive/html/bug-glpk/2012-01/threads.html)

and

  [GLPK/Known Issues](http://en.wikibooks.org/wiki/GLPK/Known_issues#Unbounded_integer_variables)

where they call this behavior _essentially innate._ The workaround is to set upper bounds.

Sage's documentation should provide this information to the user, and suggest the workaround. We should also try trapping the exception, and notifying the user.


---

Attachment


---

Attachment

If no one is in any particular rush, I'll add a patch in a few days. I'd also like the patch to allow the user to decide that (s)he wants to use the `glp_simplex` solver rather than the `glp_intopt` solver, since `glp_simplex` _does not_ crash on these inputs.


---

Comment by john_perry created at 2012-01-24 22:09:27

The patch I am _about_ to upload (not the one I just finished uploading) adds the preprocessing option, an exception for preprocessing failure, a `sig_str()` and `sig_off()` pair that catch the crash on the infeasible system in the example, documentation of the new solver option, and a doctest that raises the new exception. The doctest passes on my machine.

In short, it does everything the ticket asked for. Needs review!


---

Comment by john_perry created at 2012-01-24 22:09:27

Changing status from new to needs_review.


---

Comment by john_perry created at 2012-01-24 22:50:57

I just discovered a bug in the `copy` function: I did not copy the `preprocessing` option. For that matter, ticket #10785 did not copy the `tm_lim` option that it introduced. The new patch takes care of those.


---

Comment by ncohen created at 2012-01-25 09:14:52

Hellooooooo !!

It seems weird that the loglevel has to be stored in *BOTH* iocp and smcp `O_o`
Isn't just one of them necessary ? And you also thought of copying the self.preprocessing variable but it is not the only one that can be set with GLPK. The "timelimit" option is also exposed, and we probably can do better than to store everything twice and redefine it in the new copy. Isn't there a "GLPK way" to copy the content of a scmp object ? It would be nice to let GLPK do the job, or we would have to create a new variable each time we expose a new parameter `O_o`

By the way, I reinstalled Sage and tested the CBC patch, but of course it still returns this "gzopen function non found" error message I told you some time ago... And as the fix I had found does not work on your computer we are not done yet `:-D`

Nathann


---

Comment by john_perry created at 2012-01-25 16:34:02

Replying to [comment:5 ncohen]:

> It seems weird that the loglevel has to be stored in *BOTH* iocp and smcp `O_o` Isn't just one of them necessary ?

`glp_simplex` doesn't take `iocp` parameters.

> Isn't there a "GLPK way" to copy the content of a scmp object ? It would be nice to let GLPK do the job, or we would have to create a new variable each time we expose a new parameter

Even if true, and it probably is, that would still be overkill, given our current setup. The `smcp` and `iocp` structures are created in `__cinit__` regardless of whether we're copying or creating. If the `copy` function were then to invoke GLPK's copy constructor (or whatever) then we would be creating the same thing twice, merely to avoid setting two parameters directly?


---

Comment by ncohen created at 2012-01-26 09:56:53

> `glp_simplex` doesn't take `iocp` parameters.

Right right, so you need to update its log level too ! Sorry about that.

> Even if true, and it probably is, that would still be overkill

Granted, granted. So I only have two remarks on your patch :
    * First, I guess that ``self.scmp`` behaves like ``self.iocp`` and should also be deallocated in ``__dealloc__``
    * Could you add a doctest that shows the signal returned by your code when a bad lp is solved without the preprocessing flag, and how it is solved afterwards ? The shortest bad example I was able to produce from you .py file is this one :

```
lp = MixedIntegerLinearProgram(solver="GLPK")
lp.add_constraint( lp[1] +lp[2] -2.0 *lp[3] , max =-1.0)
lp.add_constraint( lp[0] -4.0/3 *lp[1] +1.0/3 *lp[2] , max =-1.0/3) # <-- succeeds up to here w/last line                                                                                                                                                                        
lp.add_constraint( lp[0] +0.5 *lp[1] -0.5 *lp[2] +0.25 *lp[3] , max =-0.25) # <--- fails here w/last line                                                                                                                                                                        
lp.add_constraint( lp[0] +4.0 *lp[1] -lp[2] +lp[3] , max =-1.0) # <-- last line                                                                                                                                                                                                  
lp.solve()
```


I also believe the documentation you added about the ``preprocessing`` flag is worth being in a .. WARNING:: environment. This way it'll appear in red in the HTML documentation, and should attract the attention of somebody having some troubles with GLPK. That's up to you `:-)`

Sorry if I am being a little slow, there are several nasty patches needing attention at the same time (not to mention the awful CBC one) and "free time" is a long-forgotten dream these days `O_o`

Nathann


---

Comment by john_perry created at 2012-01-26 16:23:31

Replying to [comment:7 ncohen]:
>     * First, I guess that ``self.scmp`` behaves like ``self.iocp`` and should also be deallocated in ``__dealloc__``

Duh. :-)

>     * Could you add a doctest that shows the signal returned by your code when a bad lp is solved without the preprocessing flag, and how it is solved afterwards ?

Wow, thanks! I didn't think I could get one that small easily. I will replace the doctest I added with that one instead.

> I also believe the documentation you added about the ``preprocessing`` flag is worth being in a .. WARNING:: environment. This way it'll appear in red in the HTML documentation, and should attract the attention of somebody having some troubles with GLPK. That's up to you `:-)`

That's perfect; I wasn't really aware of it.

> Sorry if I am being a little slow, there are several nasty patches needing attention at the same time (not to mention the awful CBC one) and "free time" is a long-forgotten dream these days `O_o`

I know what you mean, and I'm not complaining. If a couple of weeks passed, I'd more likely post a request at sage-devel anyway.

Thanks a lot!


---

Comment by john_perry created at 2012-01-26 16:48:13

Replying to [comment:7 ncohen]:
>     * Could you add a doctest that shows the signal returned by your code when a bad lp is solved without the preprocessing flag, and how it is solved afterwards ?

Apparently not, by the way. If I try to solve without setting the parameter, the doctest fails because an exception is raised.

I'm looking at the manual to see if I can find a way around this.


---

Comment by ncohen created at 2012-01-26 16:50:15

> Apparently not, by the way. If I try to solve without setting the parameter, the doctest fails because an exception is raised.

Oh, but you can "test" exceptions too ! Look at a file like mip.pyx and look for "traceback"

Nathann


---

Comment by john_perry created at 2012-01-26 16:56:27

Replying to [comment:10 ncohen]:
> Oh, but you can "test" exceptions too ! Look at a file like mip.pyx and look for "traceback"

Hey, you're awake :-) Thanks, that worked. New patch in a sec.


---

Comment by john_perry created at 2012-01-26 17:01:21

Wait. Is the warning supposed to be _indented_ inside the `..WARNING::` ?


---

Comment by john_perry created at 2012-01-26 17:04:50

Oops. Yes, it _is_ supposed to be indented. Hopefully, that's fixed now.


---

Attachment

Helloooooo !!

Ahem. Your patch was missing a space between ".." and "warning", and because of that the documentation was not properly built. I could not bring myself to set the ticket back to "needs work" so I changed it myself and reuploaded your own patch after adding the space `^^;`

Good to go ! `:-)`


---

Comment by ncohen created at 2012-01-26 18:30:02

Changing status from needs_review to positive_review.


---

Comment by john_perry created at 2012-01-26 18:35:49

Replying to [comment:14 ncohen]:
> Ahem. Your patch was missing a space between ".." and "warning", and because of that the documentation was not properly built. I could not bring myself to set the ticket back to "needs work" so I changed it myself and reuploaded your own patch after adding the space `^^;`

LOL

How do I test that the documentation is built properly? Run it in the notebook?


---

Comment by ncohen created at 2012-01-27 08:52:08

Yoooooooooooo !!

> How do I test that the documentation is built properly? Run it in the notebook?

nop ! You just have to run "sage -docbuild reference html"
The first run may (it depends on the releases) take some times, but you can rebuild it in a few seconds afterwards when you update the files.

It will tell you where the html files are located on your computer. And I guess it will also update the one used by the notebook though I really never use it `^^;`

Nathann


---

Comment by jdemeyer created at 2012-02-02 12:54:24

Resolution: fixed
